bayes_kit/ensemble.py

```diff
@@ -1,68 +1,91 @@
+        if self._thetas.shape != self._drawshape:
+            raise ValueError(f"init must be shape of draw {self._drawshape}; found {self._thetas.shape=}")
+        self._firsthalf = range(0, int(self._halfwalkers))
+        self._secondhalf = range(int(self._halfwalkers), int(self._walkers))
 
-    # def __iter__(self):
-    #     return self
+    def __iter__(self) -> Iterator[Sample]:
+        return self
 
-    # def __next__(self):
-    #     return self.sample
+    def __next__(self) -> Sample:
+        return self.sample()
 
-    # def draw_z(self):
-    #     """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
-    #     return np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a))
+    def draw_z(self) -> Sample:
+        """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
+        return np.asarray(np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a)))
 
-    # def stretch_move(self, theta_k: NDarray[np.float64], theta_j: NDarray[np.float64]):
-    #     z = self.draw_z()
-    #     theta_star = theta_j + z * (theta_k - theta_j)  # (1 - z) * theta_j + z * theta_k
-    #     log_q = (self._dims - 1) * np.log(z) + self._model.log_density(theta_star) - self._model.log_density(theta_k)
-    #     if np.log(np.random.uniform()) < log_q:
-    #         return theta_star
-    #     return theta_k
+    def stretch_move(self, theta_k: NDArray[np.float64], theta_j: NDArray[np.float64]) -> Sample:
+        z = self.draw_z()
+        theta_star = np.asarray(theta_j + z * (theta_k - theta_j))  # (1 - z) * theta_j + z * theta_k
+        print(f"{theta_k=}  {theta_j=}  {z=}  {theta_star=}")
+        log_q = (self._dim - 1) * np.log(z) + self._model.log_density(theta_star) - self._model.log_density(theta_k)
+        log_u = np.log(np.random.uniform())
+        print(f"{log_q=}  {log_u=}")
+        if log_u < log_q:
+            return theta_star
+        return theta_k
 
-    # def sample(self) -> NDarray[np.float64]
-    #     js = np.random.choice(secondhalf, size=self._halfwalkers)
-    #     for k in firsthalf:
-    #         self._thetas[k] = stretch_move(self._thetas[k], self._thetas[js[k]])
-    #     js = np.random.choice(firsthalf, size=self._halfwalkers)
-    #     for k in secondhalf:
-    #         self_thetas[k] = stretch_move(self._thetas[k], self._thetas[js[k]])
-    #     return self._thetas
+    def sample(self) -> Sample:
+        print(f"IN: {self._thetas=}")
+        js = np.random.choice(self._secondhalf, size=self._halfwalkers, replace=False)
+        for k in self._firsthalf:
+            self._thetas[k] = self.stretch_move(self._thetas[k], self._thetas[js[k]])
+        js = np.random.choice(self._firsthalf, size=self._halfwalkers, replace=False)
+        for k in self._secondhalf:
+            self._thetas[k] = self.stretch_move(self._thetas[k], self._thetas[js[k - self._halfwalkers]])
+        print(f"OUT: {self._thetas=}")
```

Comment:
There are a handful of print statements which seem like they should be removed before this is checked in.

We can talk about a debug logging strategy if you'd like some of this information to be preserved

Reply:
Thanks.  I don't need to save them, I had just forgotten to remove before creating the PR.